### PR TITLE
About and Print inform about projections

### PR DESCRIPTION
**Pull request by herbelin**
The PR adds the following:
- if `x` is a projection, `About x` indicates it
- if `x` is a projection, `Print x` prints the whole record to which this projection refers, as it would do for the constructor of the record
- for `Record`, prints the name of the `as` variable (so that the names used in `Arguments` make sense)

Examples:
```coq
Set Primitive Projections.
Record prod (A B : Type) : Type := Build_prod { fst : A; snd : B }.
About fst.
(*
fst : forall A B : Type, prod A B -> A

fst is not universe polymorphic
fst is a primitive projection of prod <-------- new information
Arguments fst (A B)%type_scope p
fst is transparent
Expands to: Top.fst
*)
```
```coq
Record prod (A B : Type) : Type := Build_prod { fst : A; snd : B }.
About fst.
(*
fst : forall A B : Type, prod A B -> A

fst is not universe polymorphic
fst is a projection of prod <-------- new information
Arguments fst (A B)%type_scope p
fst is transparent
Expands to: Top.fst
*)
```
```coq
Set Primitive Projections.
Record prod (A B : Type) : Type := Build_prod { fst : A; snd : B }.
Print fst.
(*
Record prod (A B : Type) : Type := Build_prod { fst : A;  snd : B } as p.

prod has primitive projections with eta conversion.
Arguments prod (A B)%type_scope
Arguments Build_prod (A B)%type_scope fst snd
Arguments fst (A B)%type_scope p
Arguments snd (A B)%type_scope p
*)
(* Before:
fst = fun (A B : Type) (p : prod A B) => fst _ _ p
     : forall A B : Type, prod A B -> A

Arguments fst (A B)%type_scope p
*)
```

- [x] Added / updated **test-suite**.
- [ ] Added **changelog**.

### CODE REVIEW

File: test-suite/output/Projections.out [hunk 1]
```diff
@@ -1,17 +1,20 @@
 fun S : store => S.(store_funcs)
      : store -> host_func
-a =
-fun A : Type =>
-let B := A in fun (C : Type) (u : U A C) => (A, B, C, c _ _ u)
-     : forall A : Type,
-       let B := A in
-       forall C : Type, U A C -> Type * Type * Type * (B * A * C)
+Record U (A : Type) (B : Type := A) (C : Type) : Type := Build_U
+  { c : (A * A * C)%type;  a := (A, A, C, c);  b : a = a } as u.
```
Comment (SkySkimmer) [inline test-suite/output/Projections.out:4]: It seems the B uses got unfolded (declaration is `{c:B*A*C;a:=(A,B,C,c);b:a=a}`), is it hard to keep them?

**Review by SkySkimmer — DISMISSED** [review]
I don't think we should print the `as` when it is the automatic one, we should print it only when the user specified a different one.

File: test-suite/output/Projections.out (hunk 1 above)
Comment (herbelin) [reply test-suite/output/Projections.out:4]: Interesting! Probably it was already unfolded in the type of `Build_U` in kernel. Will have to be checked.

**Comment by herbelin** [comment]
> I don't think we should print the as when it is the automatic one, we should print it only when the user specified a different one.

There are pros and cons, I don't know. In any case, technically, I don't think we remember the information of whether the name was "canonically" generated or manually given.

**Comment by SkySkimmer** [comment]
It doesn't matter if it was manually given, it only matters if it's different from the generated one. Same way we don't print `return` for `match` when it's trivial etc.

**Comment by herbelin** [comment]
> If it's different from the generated one

OK, then we could test if the name is the same as the generated one and not print it if it is so. Should work.

**Comment by herbelin** [comment]
@skyskimmer: the `as` is now printed only when canonical!

File: test-suite/output/Projections.out (hunk 1 above)
Comment (herbelin) [reply test-suite/output/Projections.out]: It was indeed (only) a printing bug, so relevant for the PR. Fixed.

**Comment by SkySkimmer** [comment]
Do we care that Prnt on a projection now does not show the definition of the projection? Especially for non primitive projections it means the primitive term doesn't get shown.

**Comment by herbelin** [comment]
> Do we care that Prnt on a projection now does not show the definition of the projection? Especially for non primitive projections it means the primitive term doesn't get shown.

Yes, I wondered about it and it is disputable. We can argue that it can be reconstructed, the same way as we don't display the type of a constructor on the basis that it can be reconstructed.

Also, if one really wants the type of a projection constant `p`, we can do `Check @p`.

**Comment by SkySkimmer** [comment]
`About p` still shows its type not the record AFAICT (same for About of a constructor).

It's only the body for Print that I'm concerned about.

**Comment by herbelin** [comment]
> but that gets verbose.

I tried on the examples of the test suite and it seems indeed a bit verbose. E.g.:
```coq
Record RWrap@{uu} (A : Type@{uu}) : Type@{uu} := rwrap
  { runwrap : A }.
(* uu |=  *)

runwrap@{uu} =
fun (A : Type@{uu}) (r : RWrap@{uu} A) => let (runwrap) := r in runwrap
     : forall A : Type@{uu}, RWrap@{uu} A -> A
(* uu |=  *)

Arguments RWrap A%type_scope
Arguments rwrap A%type_scope runwrap
Arguments runwrap A%type_scope r
```
If ok to you, I would rather print only the body of non-primitive projections, but accompanied with a link to the type so that one can further do `Print` on the type.

**Comment by SkySkimmer** [comment]
Sounds good

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**